File: android/core-model/src/main/kotlin/com/rfmapper/core/model/ExportManifest.kt
package com.rfmapper.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class PackageType {
    OBSERVATIONS,
    DERIVED,
}

@Serializable
enum class ExportKind {
    DAY,
    SESSION,
    RANGE,
}

@Serializable
data class DateRange(
    @SerialName("from") val from: String,
    @SerialName("to") val to: String,
)

@Serializable
data class Generator(
    @SerialName("name") val name: String,
    @SerialName("version") val version: String,
)

/**
 * `manifest.json` for an observation package.
 *
 * [countsBySensorType] and [observationCount] let the Master's import preview summarise a package
 * without parsing 18,000 rows, and let a mismatch against the actual content be detected as
 * `MANIFEST_COUNT_MISMATCH` rather than silently accepted.
 */
@Serializable
data class ExportManifest(
    @SerialName("schema_version") val schemaVersion: String = SchemaVersion.CURRENT,
    @SerialName("package_type") val packageType: PackageType = PackageType.OBSERVATIONS,
    @SerialName("export_id") val exportId: String,
    @SerialName("observer_id") val observerId: String,
    @SerialName("created_at") val createdAt: String,
    @SerialName("export_kind") val exportKind: ExportKind,
    @SerialName("date_range") val dateRange: DateRange? = null,
    @SerialName("observation_count") val observationCount: Long,
    @SerialName("first_observation") val firstObservation: String? = null,
    @SerialName("last_observation") val lastObservation: String? = null,
    @SerialName("app_version") val appVersion: String,
    @SerialName("platform") val platform: String? = null,
    @SerialName("os_version") val osVersion: String? = null,
    @SerialName("device_model") val deviceModel: String? = null,
    @SerialName("installation_id") val installationId: String? = null,
    @SerialName("session_ids") val sessionIds: List<String> = emptyList(),
    @SerialName("counts_by_sensor_type") val countsBySensorType: Map<String, Long> = emptyMap(),
    @SerialName("ground_truth_count") val groundTruthCount: Long = 0,
    @SerialName("generator") val generator: Generator,
) {
    init {
        require(observerId.isNotBlank()) { "observer_id must not be blank" }
        require(observationCount >= 0) { "observation_count must not be negative" }
        if (observationCount > 0) {
            require(firstObservation != null && lastObservation != null) {
                "a non-empty package must declare first_observation and last_observation"
            }
        }
    }
}

/** `manifest.json` for a derived package produced by the Positioning Lab. */
@Serializable
data class DerivedManifest(
    @SerialName("schema_version") val schemaVersion: String = SchemaVersion.CURRENT,
    @SerialName("package_type") val packageType: PackageType = PackageType.DERIVED,
    @SerialName("export_id") val exportId: String,
    @SerialName("created_at") val createdAt: String,
    @SerialName("date_range") val dateRange: DateRange? = null,
    @SerialName("algorithm_version") val algorithmVersion: String,
    @SerialName("engine_versions") val engineVersions: Map<String, String>,
    @SerialName("parameter_set_sha256") val parameterSetSha256: String? = null,
    @SerialName("random_seed") val randomSeed: Int? = null,

    /**
     * The import batches this output was computed from. The Master rejects a derived package that
     * references data it has never seen, which is what keeps the derived layer traceable back to
     * raw evidence.
     */
    @SerialName("source_dataset_ids") val sourceDatasetIds: List<String>,
    @SerialName("reference_model_id") val referenceModelId: String? = null,
    @SerialName("calibration_set_id") val calibrationSetId: String? = null,

    /** SYNTHETIC results prove pipeline correctness only; they must never be shown as site accuracy. */
    @SerialName("dataset_kind") val datasetKind: DatasetKind = DatasetKind.REAL,
    @SerialName("counts") val counts: Map<String, Long> = emptyMap(),
    @SerialName("generator") val generator: Generator,
)

@Serializable
enum class DatasetKind {
    REAL,
    SYNTHETIC,
    MIXED,
}

File: ios/Packages/RFMapper/Sources/RFMapperCore/ExportManifest.swift
/// `manifest.json` for an observation package. Mirrors `core-model/ExportManifest.kt`.
///
/// `countsBySensorType` and `observationCount` let the Master's import preview summarise a package
/// without parsing 18,000 rows, and let a mismatch against the actual content be detected as
/// `MANIFEST_COUNT_MISMATCH` rather than silently accepted.
public struct ExportManifest: Equatable, Sendable {

    public var schemaVersion: String
    public var packageType: PackageType
    public var exportId: String
    public var observerId: String
    public var createdAt: String
    public var exportKind: ExportKind
    public var dateRange: DateRange?
    public var observationCount: Int64
    public var firstObservation: String?
    public var lastObservation: String?
    public var appVersion: String
    public var platform: String?
    public var osVersion: String?
    public var deviceModel: String?
    public var installationId: String?
    public var sessionIds: [String]
    public var countsBySensorType: [String: Int64]
    public var groundTruthCount: Int64
    public var generator: Generator

    public init(
        schemaVersion: String = SchemaVersion.current,
        packageType: PackageType = .observations,
        exportId: String,
        observerId: String,
        createdAt: String,
        exportKind: ExportKind,
        dateRange: DateRange? = nil,
        observationCount: Int64,
        firstObservation: String? = nil,
        lastObservation: String? = nil,
        appVersion: String,
        platform: String? = nil,
        osVersion: String? = nil,
        deviceModel: String? = nil,
        installationId: String? = nil,
        sessionIds: [String] = [],
        countsBySensorType: [String: Int64] = [:],
        groundTruthCount: Int64 = 0,
        generator: Generator
    ) {
        self.schemaVersion = schemaVersion
        self.packageType = packageType
        self.exportId = exportId
        self.observerId = observerId
        self.createdAt = createdAt
        self.exportKind = exportKind
        self.dateRange = dateRange
        self.observationCount = observationCount
        self.firstObservation = firstObservation
        self.lastObservation = lastObservation
        self.appVersion = appVersion
        self.platform = platform
        self.osVersion = osVersion
        self.deviceModel = deviceModel
        self.installationId = installationId
        self.sessionIds = sessionIds
        self.countsBySensorType = countsBySensorType
        self.groundTruthCount = groundTruthCount
        self.generator = generator
    }

    public func validate() throws {
        var reasons: [String] = []
        if observerId.trimmingASCIIWhitespace().isEmpty {
            reasons.append("observer_id must not be blank")
        }
        if observationCount < 0 {
            reasons.append("observation_count must not be negative")
        }
        if observationCount > 0 && (firstObservation == nil || lastObservation == nil) {
            reasons.append("a non-empty package must declare first_observation and last_observation")
        }
        guard reasons.isEmpty else {
            throw ObservationError.invalid(id: exportId, reasons: reasons)
        }
    }

    /// Declaration order matches the Kotlin data class, because the bytes are part of the contract.
    public var json: JSONValue {
        .object([
            ("schema_version", .string(schemaVersion)),
            ("package_type", .string(packageType.rawValue)),
            ("export_id", .string(exportId)),
            ("observer_id", .string(observerId)),
            ("created_at", .string(createdAt)),
            ("export_kind", .string(exportKind.rawValue)),
            ("date_range", dateRange?.json ?? .null),
            ("observation_count", .int(observationCount)),
            ("first_observation", .string(firstObservation)),
            ("last_observation", .string(lastObservation)),
            ("app_version", .string(appVersion)),
            ("platform", .string(platform)),
            ("os_version", .string(osVersion)),
            ("device_model", .string(deviceModel)),
            ("installation_id", .string(installationId)),
            ("session_ids", .array(sessionIds.map { .string($0) })),
            ("counts_by_sensor_type", .sortedCounts(countsBySensorType)),
            ("ground_truth_count", .int(groundTruthCount)),
            ("generator", generator.json),
        ])
    }
}

public struct DateRange: Equatable, Sendable {
    public var from: String
    public var to: String

    public init(from: String, to: String) {
        self.from = from
        self.to = to
    }

    public var json: JSONValue {
        .object([("from", .string(from)), ("to", .string(to))])
    }
}

public struct Generator: Equatable, Sendable {
    public var name: String
    public var version: String

    public init(name: String, version: String) {
        self.name = name
        self.version = version
    }

    public var json: JSONValue {
        .object([("name", .string(name)), ("version", .string(version))])
    }
}
